middleware.ts

```diff
@@ -131,9 +225,104 @@ export async function middleware(req: NextRequest) {
 
-  // From layer 3
   // Set cookie on response back to browser so client can render correct language on client components
-  if (cookieSyncRequired) response.cookies.set("i18next", pathLang);
+  if (pathLang && cookieLang !== pathLang) response.cookies.set("i18next", pathLang);
 
   return response;
-}
+};
+
+/**
+ * Handles the redirection of users in the authentication flow
+ */
+const authFlowRedirect = (
+  req: NextAuthRequest,
+  pathname: string,
+  pathLang: string | undefined,
+  cookieLang: string | undefined
+) => {
+  const path = pathname.replace(`/${pathLang}/`, "/");
+
+  const lang = pathLang || cookieLang || fallbackLng;
+
+  const onAuthFlow = path.startsWith("/auth/mfa") || path.startsWith("/auth/restricted-access");
+
+  const onSupport =
+    path.startsWith("/support") || path.startsWith("/sla") || path.startsWith("/terms-of-use");
+
+  const session = req.auth;
+
+  const origin = req.nextUrl.origin;
+
+  // Ignore if user is in the auth flow of MfA
+  if (session && !onAuthFlow) {
+    if (
+      !session.user.hasSecurityQuestions &&
+      !path.startsWith("/auth/setup-security-questions") &&
+      // Let them access support related pages if having issues with Security Questions
+      !onSupport
+    ) {
+      logMessage.debug(
+        "Middlware Action: User has not setup security questions, redirecting to setup-security-questions"
+      );
+      // check if user has setup security questions setup
+
+      const securityQuestionsPage = new URL(`/${lang}/auth/setup-security-questions`, origin);
+      logMessage.debug(`Middleware: Redirecting to ${securityQuestionsPage}`);
+      return NextResponse.redirect(securityQuestionsPage);
+    }
+    // Redirect to policy page only if users aren't on the policy, support, or security questions page
+    if (
+      session.user.hasSecurityQuestions &&
+      !session.user.acceptableUse &&
+      !onSupport &&
+      !path.startsWith("/auth/policy") &&
+      // If they don't want to accept let them log out
+      !path.startsWith("/auth/logout")
+    ) {
+      logMessage.debug(
+        "Middleware Action: User has not accepted the Acceptable Use Policy, redirecting to policy"
+      );
+      // If they haven't agreed to Acceptable Use redirect to policy page for acceptance
+      // Also check that the path is local and not an external URL
+
+      const acceptableUsePage = new URL(
+        `/${lang}/auth/policy?referer=/${lang}${localPathRegEx.test(path) ? path : "/forms"}`,
+        origin
+      );
+      return NextResponse.redirect(acceptableUsePage);
+    }
+  }
+};
+
+/**
+ * Ensures that a user visiting a page that requires authentication
+ * is redirected to the login page if they are not authenticated
+ */
+const pageRequiresAuth = (req: NextAuthRequest, pathname: string, pathLang: string) => {
+  const path = pathname.replace(`/${pathLang}/`, "/");
+  const session = req.auth;
+
+  const pathsRequiringAuth = [
```

Comment:
Do we need to add any form builder paths here i.e. settings & responses ?  

Just asking haven't tested yet.

Reply:
The settings and responses pages in form builder both have unauthenticated views so we can't protect them in middleware and will need to continue to do the check in the pages/components themselves.